### ISSUE TITLE
Feature request: add parent and root reference to assert function arguments

### ISSUE THREAD

**Issue by joseandrespg** [sylvainpolletvillard/ObjectModel#39]
Hi, 

This is a nice feature that will be nice to have to improve our assertions.

The use case is when a validation needs context from a parent model.

Currently, the only available argument in a assert function is the model itself. But having the ability to access the parent and the root object will allow us to create more precise error messages (having the ability of create an validation error in the right path of the object and not in the parent/root).

Basic example of how it would work:

```javascript
var Player = new Model({
    name: String,
    salary: Number
});

var Team = new Model({
    name: String,
    titles: Number,
    players: Model.Array(Player)
});

var Season = new Model({
    year: Number,
    teams: Model.Array(Team)
});

/**
 * @var player current model
 * @var parent Team model
 * @var root Season model
 */
Player.assert(function hasTeamRightSalary(player, parent, root) {
  if (parent.titles > 10) {
      return player.salary > 100000;
  }

  return true;
});

Player.assert(function hasSeasonRightSalary(player, parent, root) {
  if (root.year > 2016) {
      return player.salary < 200000
  }

  return true;
});
```

What are your thoughts about it?

Thanks!

**Comment by sylvainpolletvillard** [comment]
Hi,

To be clear, these `parent` and `root` references will not point to the same instances depending on the context of the initial mutation. Mutation on a parent model triggers validation on this model, and consequently on all sub-models, but this validation does not "bubble up" to identify parent models impacted.

Example:
```
var player = new Player({ name: "Jordan", salary: 50000 })
var team = new Team({ name: "Chicago", titles: 3, players: [player] })
var season =  new Season({ year: 2006, teams: [team] })

Player.assert(function debug(player, parent, root) { console.log({ root, parent }); });

player.salary = 100000; // root === player ; parent === null ?
team.players.push(player) // root === team ; parent === team
season.teams.push(team) // root === season ; parent = team
```

So these references are quite unpredictable. `parent` can be null or can be any other model where Player is used, may it be an Array model, a Function model, an Object model...

My concern is: how are you going to use this if you can not predict to what model they refer ?

**Comment by joseandrespg** [comment]
In our context, it will be always predictable. Our models are submodels inside a big one (NoSQL Database).

So the user input would be like: 

```json
{
    "year": 2006,
    "teams": [
        {
            "name": "Chicago",
            "titles": 3,
            "players": [
                {
                    "name": "Jordan", 
                    "salary": 50000
                }
            ]
        }
    ]
}
``` 

We split this big input into the different submodels, so we can add the validations right to the property we want to test (apart from code maintainability).

For the case that you comment, edit an existing object, its true, you will need to test if the parent is present to perform the validation or not. It wouldn't be invalid until you create the relationship.

And in the case that a model could be used in more than one place, you may need to extend a base model to apply custom validations to those who need it. Or going on further changes add the ability to set the name of the model in the constructor, then you would be able to test what model is the parent like:

```javascript
   var player = new Model({...}, 'Player');
   player.constructor.name === 'Player';
``` 

Does it make sense?

**Comment by sylvainpolletvillard** [comment]
yes or more simply `player.constructor === Player` ; see "How can I get the model from an instance ?" from http://objectmodel.js.org/#common-questions

Even if you have a single model structure and do not reuse models between different structures, I doubt it will be as much predictable as you pretend. Regarding this example:
```
Player.assert(function hasSeasonRightSalary(player, parent, root) {
  if (root.year > 2016) {
      return player.salary < 200000
  }

  return true;
});
```

If `root` represent the initial object mutated, it can be a Player, a Team or a Season. So to fix the test you  would have to write `if(root instanceof Season && root.year > 2016)`. 

Now when a Season object is mutated, i.e `season.year = 2017`, the mutation will propagate to Player instances and check if they are valid. Suppose `player` is still valid because salary is under 200000. Then, if we mutate player salary : `player.salary = 300000` ; the mutation occures on a Player instance, so `root` will be `player`. The assertion will pass, but of course it's not what you would expect.

One way to solve this problem would be to use a circular reference  ; for example, add a `season` property in the `Player` model and test it instead of `root`. See "How should I deal with circular references in my model definitions ?" in common questions. This way, any mutation will check the expected reference. But circular references have caveats : they can't be serialized and can produce infinite loops if you don't pay attention.

Anyway I think the root of the problem is this interdependence between models. Mutation events propagation has to go in only one direction, otherwise there is always a risk of infinite loops ; basically `A.assert(validationDependingOnB)` and `B.assert(validationDependingOnA)`

So in short, I do not recommend to write models that depends on a parent context. Remember this is type checking: if you log a model instance to the console you should be able to tell if this instance is valid or not, without having to ask for information outside its scope.

**Comment by joseandrespg** [comment]
Closing. We moved this "complex" assertions out of the objectmodel definition.
